sound/soc/sof/topology.c

```diff
@@ -2477,34 +2482,80 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_pdata *sof_pdata = sdev->pdata;
 	const struct firmware *fw;
+	const char **tplg_files;
+	int tplg_cnt = 0;
 	int ret;
+	int i;
 
 	dev_dbg(scomp->dev, "loading topology:%s\n", file);
 
-	ret = request_firmware(&fw, file, scomp->dev);
-	if (ret < 0) {
-		dev_err(scomp->dev, "error: tplg request firmware %s failed err: %d\n",
-			file, ret);
-		dev_err(scomp->dev,
-			"you may need to download the firmware from https://github.com/thesofproject/sof-bin/\n");
-		return ret;
+	tplg_files = kcalloc(scomp->card->num_links, sizeof(char *), GFP_KERNEL);
+	if (!tplg_files)
+		return -ENOMEM;
+
+	if (sof_pdata->machine->get_tplg_files) {
+		tplg_cnt = sof_pdata->machine->get_tplg_files(scomp->card, sof_pdata->machine,
+							      sof_pdata->tplg_filename_prefix,
+							      &tplg_files);
+		if (tplg_cnt < 0)
+			return tplg_cnt;
 	}
 
-	if (sdev->dspless_mode_selected)
-		ret = snd_soc_tplg_component_load(scomp, &sof_dspless_tplg_ops, fw);
-	else
-		ret = snd_soc_tplg_component_load(scomp, &sof_tplg_ops, fw);
+	if (!tplg_cnt) {
+		tplg_files[0] = file;
+		tplg_cnt = 1;
+	}
 
-	if (ret < 0)
-		dev_err(scomp->dev, "error: tplg component load failed %d\n",
-			ret);
+	for (i = 0; i < tplg_cnt; i++) {
+		dev_dbg(scomp->dev, "Requesting tplg_files[%d] %s\n", i, tplg_files[i]);
+		ret = firmware_request_nowarn(&fw, tplg_files[i], scomp->dev);
+		if (ret < 0) {
+			if (i == 0 && tplg_cnt > 1) {
+				dev_dbg(scomp->dev, "Fail back to %s\n", file);
+				tplg_cnt = 1;
+				ret = request_firmware(&fw, file, scomp->dev);
+				if (ret < 0) {
+					dev_err(scomp->dev, "error: tplg request firmware %s failed err: %d\n",
+						file, ret);
+					dev_err(scomp->dev,
+						"you may need to download the firmware from https://github.com/thesofproject/sof-bin/\n");
+					goto out;
+				}
+			} else {
+				if (i > 0)
+					snd_soc_tplg_component_remove(scomp);
```

Comment:
actually after thinking about it, I think you dont need to do this at all. All of the aub-topologies are associated with the same component, so if a failure happens during loading for any of the sub-topologies, the component remove will handle all of the previously successfully loaded objects automatically. Maybe you can just add that comment here for clarity